x/tokenomics/keeper/settle_session_accounting.go

```diff
@@ -2,15 +2,138 @@ package keeper
+		return types.ErrTokenomicsSessionHeaderNil
+	}
+
+	// Validate the session header
+	if err := sessionHeader.ValidateBasic(); err != nil {
+		logger.Error("received an invalid session header", "error", err)
+		return types.ErrTokenomicsSessionHeaderInvalid
+	}
+
+	// Decompose the claim into its constituent parts for readability
+	supplierAddress, err := sdk.AccAddressFromBech32(claim.SupplierAddress)
+	if err != nil {
+		return types.ErrTokenomicsSupplierAddressInvalid
+	}
+	applicationAddress, err := sdk.AccAddressFromBech32(claim.SessionHeader.ApplicationAddress)
+	if err != nil {
+		return types.ErrTokenomicsApplicationAddressInvalid
+	}
+	root := (smt.MerkleRoot)(claim.RootHash)
+
+	// Retrieve the application
+	application, found := k.appKeeper.GetApplication(ctx, applicationAddress.String())
+	if !found {
+		logger.Error(fmt.Sprintf("application for claim with address %s not found", applicationAddress))
+		return types.ErrTokenomicsApplicationNotFound
+	}
+
+	// Retrieve the sum of the root as a proxy into the amount of work done
+	claimComputeUnits := root.Sum()
+
+	logger.Info(fmt.Sprintf("About to start settling claim for %d compute units", claimComputeUnits))
+
+	// Retrieve the existing tokenomics params
+	params := k.GetParams(ctx)
+
+	// Calculate the amount of tokens to mint & burn
+	upokt := sdk.NewInt(int64(claimComputeUnits * params.ComputeUnitsToTokensMultiplier))
+	upoktCoin := sdk.NewCoin("upokt", upokt)
+	upoktCoins := sdk.NewCoins(upoktCoin)
+
+	logger.Info(fmt.Sprintf("%d compute units equate to %d uPOKT for session %s", claimComputeUnits, upokt, sessionHeader.SessionId))
+
+	// NB: We are doing a mint & burn + transfer, instead of a simple transfer
+	// of funds from the supplier to the application in order to enable second
+	// order economic effects with more optionality. This could include funds
+	// going to pnf, delegators, enabling bonuses/rebates, etc...
+
+	// Mint uPOKT to the supplier module account
+	if err := k.bankKeeper.MintCoins(ctx, suppliertypes.ModuleName, upoktCoins); err != nil {
+		return types.ErrTokenomicsApplicationModuleFeeFailed
+	}
+
+	logger.Info(fmt.Sprintf("minted %d uPOKT in the supplier module", upokt))
+
+	// Sent the minted coins to the supplier
+	if err := k.bankKeeper.SendCoinsFromModuleToAccount(
+		ctx, suppliertypes.ModuleName, supplierAddress, upoktCoins,
+	); err != nil {
+		return types.ErrTokenomicsApplicationModuleFeeFailed
+	}
+
+	logger.Info(fmt.Sprintf("sent %d uPOKT to supplier with address %s", upokt, supplierAddress))
+
+	// Verify that the application has enough uPOKT to pay for the services it consumed
+	if application.Stake.IsLT(upoktCoin) {
+		logger.Error(fmt.Sprintf("THIS SHOULD NOT HAPPEN. Application with address %s needs to be charged more than it has staked: %v > %v", applicationAddress, upoktCoins, application.Stake))
+		// TODO_BLOCKER(@Olshansk, @RawthiL): The application was over-serviced in the last session so it basically
```

Comment:
Just a thought but rate limiting at the supplier level to ensure the application can pay for their services moving forward could be a play - nothing actionable something to think on - @RawthiL had mentioned something like this before